kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoWindowedTable.java

```diff
@@ -0,0 +1,472 @@
+    }
+    return RemoteWriteResult.success(null);
+  }
+
+  public RemoteWriteResult<SegmentPartition> postCommit(
+      final int kafkaPartition
+  ) {
+    final var segments = kafkaPartitionToSegments.get(kafkaPartition);
+
+    for (final long segmentId : segments.segmentBatch.segmentRoll.segmentsToExpire) {
+      segments.expireSegment(new SegmentPartition(kafkaPartition, segmentId));
+    }
+    segments.segmentBatch.finalizeRoll();
+    return RemoteWriteResult.success(null);
+  }
+
+  @Override
+  public long fetchOffset(final int kafkaPartition) {
+    final WindowMetadataDoc result = metadata.find(
+        Filters.eq(WindowMetadataDoc.ID, kafkaPartition)
+    ).first();
+    if (result == null) {
+      LOG.error("Offset fetch failed due to missing metadata row for partition {}", kafkaPartition);
+      throw new IllegalStateException("No metadata row found for partition " + kafkaPartition);
+    }
+    return result.offset;
+  }
+
+  public WriteModel<WindowMetadataDoc> setOffsetAndStreamTime(
+      final int kafkaPartition,
+      final long offset
+  ) {
+    final var segments = kafkaPartitionToSegments.get(kafkaPartition);
+    final long epoch = segments.epoch;
+    final long streamTime = segments.segmentBatch.batchStreamTime;
+
+    LOG.info("{}[{}] Updating offset to {} and streamTime to {} with epoch {}",
+             name, kafkaPartition, offset, streamTime, epoch);
+
+    return new UpdateOneModel<>(
+        Filters.and(
+            Filters.eq(WindowMetadataDoc.ID, kafkaPartition),
+            Filters.lte(WindowMetadataDoc.EPOCH, epoch)
+        ),
+        Updates.combine(
+            Updates.set(WindowMetadataDoc.OFFSET, offset),
+            Updates.set(WindowMetadataDoc.STREAM_TIME, streamTime),
+            Updates.set(WindowMetadataDoc.EPOCH, epoch)
+        )
+    );
+  }
+
+  public long epoch(final int kafkaPartition) {
+    return kafkaPartitionToSegments.get(kafkaPartition).epoch;
+  }
+
+  /**
+   * Inserts data into this table. Note that this will overwrite
+   * any existing entry in the table with the same key.
+   *
+   * @param kafkaPartition the kafka partition
+   * @param windowedKey    the windowed data key
+   * @param value          the data value
+   * @param epochMillis    the timestamp of the event
+   * @return a statement that, when executed, will insert the value
+   *         for this key and partition into the table.
+   *         Note that the key in this case is a "windowed" key, where
+   *         {@code windowedKey} includes both the record key and
+   *         the windowStart timestamp
+   */
+  @Override
+  public WriteModel<WindowDoc> insert(
+      final int kafkaPartition,
+      final WindowedKey windowedKey,
+      final byte[] value,
+      final long epochMillis
+  ) {
+    final var partitionSegments = kafkaPartitionToSegments.get(kafkaPartition);
+    partitionSegments.segmentBatch.updateStreamTime(windowedKey.windowStartMs);
```

Comment:
I think it would be better for the stream time to be tracked by the `WriteBatcher` (which is common for all stores) and passed in to `preFlush`. That way we don't have to repeat this for every store implementation, and it's not tracked as a side effect (which makes it a bit harder to follow how it's tracked)